PHX-47: Allow using org unit names in report-server analytics

kael89 opened the pull request:
### Issue #:
PHX-47

Based off [rn-83-convert-sum-to-typed-func](https://github.com/beyondessential/tupaia/tree/rn-83-convert-sum-to-typed-func) since it has the latest in `report-server`

rohan-bes commented on packages/expression-parser/src/expression-parser/ExpressionParser.js:
```diff
@@ -62,6 +83,14 @@ export class ExpressionParser {
+  readExpression = expression =>
+    this.CALCULATION_PREFIX
+      ? expression.replace(new RegExp(`^${this.CALCULATION_PREFIX}`), '')
+      : expression;
```
Hmmm while I'm not opposed to us doing this, the whole `=` at the start of expressions is really just a report-server thing for now... Indicators and Meditrak App Arithmetic questions both use expression parser, and don't append the `=` to the start to indicate it's a calculation.

To me it feels appropriate to keep all logic about this encapsulated in TransformParser until we've standardised the behaviour across all users of the ExpressionParser?

kael89 replied:
I see what you mean... I just tried to avoid having to override each method in `TransformParser`, to make them aware of the `=` symbol. Not from laziness, but trying to figure out a correct abstraction for this concept...

It seems to me that if a child class does alter how expressions (or a subset of them, eg `calculations`) are defined, then the parent class should expect and allow that... If not, then we are "violating" the contract of the parent class, and is an indication that we shouldn't do it in the first place.

For this specific case: an expression parser does depend on symbols to do its job. Thus, it seems fair to me that a specific implementation of the parser, either via inheritance or via constructor configuration, could use its symbols of choice.

rohan-bes commented on packages/expression-parser/src/expression-parser/ExpressionParser.js:
```diff
@@ -3,7 +3,24 @@
+ * @property {(s: string) => boolean} has
+ * @property {(s: string) => void} delete
+ * @property {() => void} delete
+ */
```
❤️

rohan-bes commented on packages/expression-parser/src/expression-parser/ExpressionParser.js:
```diff
@@ -205,4 +248,8 @@ export class ExpressionParser {
   }
+
+  factory(...args) {
+    return factory(...args);
```
```suggestion
    return this.math.factory(...args);
```
Thoughts on this? It will allow us to use the factory function from our existing math instance, which means if we've already imported stuff into that math instance we can use it (custom types for example)

kael89 replied:
Good call!

rohan-bes commented on packages/expression-parser/src/__tests__/expression-parser/ExpressionParser.test.js:
```diff
@@ -25,4 +25,59 @@ describe('ExpressionParser', () => {
+      const expected = ['cos', 'sin'];
+      expect(functions).toStrictEqual(expected);
+    });
+  });
```
Nice tests!!!

rohan-bes reviewed (changes requested):
Awesome stuff @kael89! Really nice work 🎉 

Few comments on naming and some code structure suggestions, but am really happy with things overall!

kael89 commented on packages/report-server/src/reportBuilder/context/buildContext.ts:
```diff
@@ -0,0 +1,71 @@
+};
+
+const contextBuilders: Record<ContextProp, ContextBuilder<ContextProp>> = {
+  orgUnitMap: buildOrgUnitMap,
```
Oh yes, good call, happy to use another name! I can use something simple eg `orgUnits`

kael89 commented on packages/report-server/src/reportBuilder/context/extractContextProps.ts:
```diff
@@ -0,0 +1,57 @@
+export const extractContextProps = (transform: unknown): ContextProp[] => {
+  if (!Array.isArray(transform)) {
+    return [];
+  }
```
Yes that's what I thought too! An approach would be:

1. Expand `this.config.transform` to an array of objects (eg by inlining aliases to objects)
2. Validate the expanded transform
3. Use the expanded transform to find all functions that use context
4. Use the expanded transform to build the output

So validation would happen first, and the rest of the functions would expected a validated & expanded `Transform` object as an input. This will make function extraction simpler, since the `transform` in its input will have a more specific schema.

We use a similar approach in `indicators` (we expand the aggregation input)

kael89 replied:
I can submit a PR implementing the above at a later stage, but yeah for now let's proceed with what we have

kael89 commented on packages/report-server/src/reportBuilder/context/extractContextProps.ts:
```diff
@@ -0,0 +1,57 @@
+        return typeof param !== 'object' ? [param] : Object.entries(param);
+      });
+    })
+    .flat(3)
```
Yeah good point... I guess there is no harm using a bigger value here (eg `5`, even 10!), another solution would be using [lodash.flattenDeep](https://lodash.com/docs/4.17.15#flattenDeep).

I'm not sure why the default implementation uses `1` as depth... I'd assume it is simple/performant to flatten with a generous depth by default (similar to how recursions use a max call stack size)

kael89 commented on packages/report-server/src/reportBuilder/context/index.ts:
```diff
@@ -0,0 +1,2 @@
+export * from './buildContext';
+export * from './types';
```
Haha I am happy to respect and use the existing code style in the package! I generally try to do that (although habit is strong)

Let me know what you prefer!

kael89 replied:
Changed them to explicit exports to keep consistency with the rest of the project

kael89 commented on packages/report-server/src/reportBuilder/transform/parser/TransformParser.ts:
```diff
@@ -131,6 +152,21 @@ export class TransformParser extends ExpressionParser {
+  }
+
   public static isExpression(expression: string) {
     return expression.startsWith('=');
```
Actually we don't need this method anymore, since `ExpressionParser` handles this - but I forgot to update the child class! Thanks for catching this

kael89 commented on packages/report-server/src/reportBuilder/transform/parser/functions/context.ts:
```diff
@@ -0,0 +1,26 @@
+ * `contextProps`: props that will be required by the function
+ * `create`: factory method that creates the function by injecting the context
+ */
+type ContextParams = {
+  contextProps: ContextProp[];
+  create: (deps: { getContext: () => Context }) => (...args: any[]) => unknown;
+};
```
I used `params` since it is used in a seemingly similar vein throughout the project: get an object (eg from `transform`/`fetch` config), and construct a function (the `transform`/`fetch` function). I guess that my bias towards "shorter names where possible" played a role too!

In retrospect, those are not quite the same thing, so happy to change that - it feels like equating oranges and apples here 😄 

About `contextDependencies`: that was the term I used initially. I also tried to inject the `context` object directly (instead of a getter for it: `getContext`), but math.js wouldn't let me. I think I quite like `contextProps` eventually, because that's what this is: a prop of the context - the fact that `context` is eventually injected as a "dependency", through a getter, is probably an implementation detail. It also reminds me a bit of `react` props which I feel is an advantage - you props are drilled down your components, and you don't care where they came from...

Happy to discuss this further though!

kael89 commented on packages/report-server/src/reportBuilder/transform/functions/excludeColumns.ts:
```diff
@@ -20,8 +21,8 @@ const paramsValidator = yup.object().shape({
 
-const excludeColumns = (rows: Row[], params: ExcludeColumnsParams): Row[] => {
-  const parser = new TransformParser(rows);
+const excludeColumns = (rows: Row[], params: ExcludeColumnsParams, context: Context): Row[] => {
```
I agree that while there are solutions to inject `context` in those functions, they will probably make the code more complicated, as it will be harder to follow where the context is coming from... And you have established a nice functional pattern here!

There is a tipping point where the complexity of a wrapper would be preferred over the complexity/clutter of an additional parameter like context, but I believe we haven't reached that stage yet 🙂

kael89 commented:
@rohan-bes I have addressed all comments apart from the following:

* `contextDependencies`: left it as `contextProps`

kael89 commented on packages/report-server/src/reportBuilder/transform/parser/TransformParser.ts:
```diff
@@ -65,14 +77,22 @@ export class TransformParser extends ExpressionParser {
+  }
 
-    return expression;
+  protected readExpression(input: unknown) {
+    return this.isExpression(input)
+      ? (input as string).replace(new RegExp(`^${this.EXPRESSION_PREFIX}`), '')
+      : input;
+  }
```
Eventually I followed your suggestion and moved most of the logic in `TransformParser` 🙂 Not doing so would result in some interesting questions which we probably cannot answer consistently at this stage:
  * What happens if something is not an expression? Do we throw an error or just use the input?
   * Similarly: is the prefix optional or not? And if it is optional, what's the point in using it?

The compromise I thought of is adding a `readExpression` method in the base class, which can be overridden to support custom expression formats. The caveat here is that future devs must remember to call this function when an expression needs to be read in `ExpressionParser`. It's not ideal, but extending all present (and future) `ExpressionParser` methods in `TransformParser` so that our custom expression parsing is used seems even worse!

Another advantage of this pattern is that the custom format specified in sub classes is not restricted to a prefix, but could be anything.

kael89 replied:
Regarding semantics, I still believe that anything that can be evaluated is an _expression_. Following that definition, we should treat all expressions equally:
* for the literal number `2`: use `=2`
* for the literal string `MyString`: use `='MyString'`, or `=MyString` if our parser can easily distinguish between variables, functions and string literals

**But** I guess we had a valid reason for not doing so - I remember that in the past we used expressions like the following in the left sides of params, which didn't read fluently:
`"'name'":  "orgUnit"`

kael89 replied:
I believe that left-side evaluation is what what makes a difference in `report-server` in relation to other languages we are (I am?) used into, and may be the reason - and justification! of more custom expression parsing.

Eg in Javascript:
```js
const valid = { result: 2 + 3 };
const invalid = { 2 + 3: 'result' };
```

kael89 commented on packages/report-server/src/reportBuilder/transform/parser/TransformParser.ts:
```diff
@@ -65,14 +77,22 @@ export class TransformParser extends ExpressionParser {
+  }
+
+  public evaluate(input: unknown) {
+    return this.isExpression(input) ? super.evaluate(input) : input;
```
The condition retains the existing functionality, which just returns the input (and avoids validation + evaluation) if the input does not start with `=`

rohan-bes commented on packages/report-server/src/__tests__/reportBuilder/transform/parser/functions.test.ts:
```diff
@@ -75,6 +75,27 @@ describe('functions', () => {
+  describe('context', () => {
+    describe('orgUnitCodeToName', () => {
+      const context = {
+        orgUnits: [
```
Totally happy with `orgUnits` for the name here.

At the same time it makes me sad... I feel we're losing the war on the 'Rename orgUnits to entities' fight. Org units is the right choice here, since the analytic has 'organisationUnit' in it, but the terminology is just so entrenched 😢

rohan-bes replied:
Also, wanna note that I love how the architecture allows us to write tests like this! Super clean 👍

kael89 replied:
Yeah I think we came up with something clean! 🙌

kael89 commented:
> @rohan-bes I have addressed all comments apart from the following:
> 
> * `contextDependencies`: left it as `contextProps`

@rohan-bes I forgot to mention: happy to discuss about this and be convinced otherwise too. I didn't do it for the reasons I mentioned in my comment, but it's not like "I won't do this" 😛

rohan-bes commented on packages/report-server/src/reportBuilder/context/buildContext.ts:
```diff
@@ -0,0 +1,67 @@
+const buildOrgUnits = async (reqContext: ReqContext, data: FetchResponse) => {
+  const orgUnitCodes = isEventResponse(data)
+    ? data.results.map(d => d.orgUnit)
+    : data.results.map(d => d.organisationUnit);
```
It really bugs me that events use `orgUnit` and analytics use `organisationUnit`! Not saying we do this now, but I wonder how hard it'd be to migrate events to be consistent...

kael89 replied:
I agree with you we should definitely do this at some point!

rohan-bes commented:
> @rohan-bes I have addressed all comments apart from the following:
contextDependencies: left it as contextProps

I'm cool with contextProps 👍

rohan-bes commented on packages/report-server/src/routes/helpers.ts:
```diff
@@ -3,8 +3,10 @@
+import { AccessPolicy } from '@tupaia/access-policy';
+import { MicroServiceRequestContext } from '@tupaia/server-boilerplate';
+
+type EntityApi = MicroServiceRequestContext['services']['entity'];
```
I don't mind this change, but am also not opposed to the previously behaviour of directly importing `EntityApi` from `api-client`? I think some degree of awareness that services are instances of various `api-client`s is fine throughout the system?

kael89 replied:
To do that, I'd typically need to declare `@tupaia/api-client` as a dependency of this package.... I changed this to fix an existing EsLint error.

What I could do is expose some `@tupaia/api-client` types to `@tupaia/server-boilerplate` consumers through the latter... I will try that

rohan-bes replied:
Ah, so typescript has something that works really nicely for that scenario. There's an `import type` command, which you can use to import a type from somewhere without declaring that package as a dependency 👍 Since the typescript stuff is all just compile time all `import type` statements get removed when code is built!

kael89 replied:
Ah that's really nice! Turns out the `import type` syntax was used in the first place, and the error was a bit different (`EntityApi` is not exported in that package)

rohan-bes commented on packages/report-server/src/reportBuilder/transform/transform.ts:
```diff
@@ -6,10 +6,20 @@
+      transform: string;
+      title?: string;
+      description?: string;
+    };
```
Not major, but I usually find that if I set up the yup validator properly I don't need to explicitly set up a type?

kael89 replied:
Great point! Will use that 👍

rohan-bes commented on packages/report-server/src/reportBuilder/transform/transform.ts:
```diff
@@ -68,13 +78,13 @@ const buildParams = (params: unknown): TransformParams => {
     title,
     name: transformStep,
-    apply: transformBuilders[transformStep](restOfTransformParams),
+    apply: transformBuilders[transformStep](restOfTransformParams, context as Context),
```
Is this `as Context` required? If it being optional is an issue perhaps lets just provide a default empty context?

kael89 replied:
That's a bit tricky. My understandin is that it is required in some non-aliased transforms, which makes it generally required in their "type". At this code path we are calling a non-aliased transform.

Happy to have a call and see if there is a better way around that!

rohan-bes replied:
I think the general approach with validation in report-server has been that the higher levels make type assertions to the lowest common denominator, and then it's up to the children to validate up to their required level. 

In this instance I guess that would mean that context can be optional in this scope, as it's not required by all. And then for the transforms that do require it there's a secondary validation check to assert it is defined?

Yup, happy to call on this if you want 👍

rohan-bes commented on packages/report-server/src/__tests__/reportBuilder/context/buildContext.test.ts:
```diff
@@ -0,0 +1,112 @@
+      const { fields } = queryOptions;
+      return fields ? entities.map(e => pick(e, fields)) : entities;
+    },
+  };
```
Really nice to see the hand written mock over using a mocking framework. Give that JS lets us do it, I'm quite partial to hand written mocks as I find them much easier to reason about and extend/debug!

kael89 replied:
I agree! Plus I am not a big fan of `jest`'s mocking api: too many options which is confusing, expects a `string` to the path of the mocked module (instead of importing it in the test)... I miss good ol' `sinon.createStubInstance`!

rohan-bes commented on packages/report-server/src/reportBuilder/transform/parser/TransformParser.ts:
```diff
@@ -65,14 +77,22 @@ export class TransformParser extends ExpressionParser {
-    if (TransformParser.isExpression(expression)) {
-      return super.evaluate(expression.substring(1));
-    }
+  public isExpression(input: unknown) {
```
I know I'm being nitpicky here, but mind if we make this `static` again? Given that this function independnant of the instance state, it makes sense for it to be static to me? It also means that we can check if a input is an expression without instantiating a `TransformParser` (we do this in `sortRows` for instance)

kael89 replied:
Now that `EXPRESSION_PREFIX` is not inherited from the base class, we could. I'll do that.

Disclaimer: I generally avoid static methods unless if there is a very specific reason to use one, since they are [hard to mock](https://testing.googleblog.com/2008/12/static-methods-are-death-to-testability.html) + don't fit right in an inheritance model. Of course, as with almost everything, there are justifiable use cases for them!

(The article I linked has a very ominous, click-baity title 😛 But I have also confirmed that static methods are hard to mock, at least with `sinon` )

kael89 replied:
Having said that, I've noticed in the past that instantiating `ExpressionParser` took much more than one would expect (something like `30ms`). In fact, they accounted for 70% of the slow performance in the initial implementation of [nested indicators](https://github.com/beyondessential/tupaia/pull/1835)!

I don't know if this is still the case, but I am flagging it since `TransformParser` is instantiated multiple times in `report-server`. If it is, a simple solution around that would be using a singleton.

rohan-bes replied:
I agree that it's good to be wary about static methods. You really need to be careful that they don't mutate state, and yes they are very hard to mock. That being said, in this instance, and in general, I do find them appropriate for places where they perform simple, functional logic, and that mocking wouldn't be appropriate.

For instance you could in theory mock `isExpression` to always return true when testing various expressions in report-server, to save yourself the trouble of putting `=` at the start of them. But I think that would be a misuse of mocking. Leaving it as is and adding the `=` improves your tests as it aligns your tests with how the code actually operates.

rohan-bes replied:
As for the performance concerns around instantiating `TransformParser`. I share them too, but I think for now it's acceptable. We insatiate at once per transform I believe, which could be up to 10 or so? Singleton pattern could work but would require a bit more overhead of cleaning up scope between uses...

kael89 replied:
10 times 30 is 300ms, a significant overhead for something that doesn't access the file system...

rohan-bes reviewed (approved):
Awesome work @kael89! Really love how this all came together. Also thanks for adding such thorough tests 🎉 

A few comments on typescript things, and a few suggestions, but nothing too major so happy to approve 👍

kael89 commented:
@rohan-bes thanks for the insightful suggestions! I've applied the changes. Let me know if you want to chat about https://github.com/beyondessential/tupaia/pull/3253#discussion_r713471400

rohan-bes commented on packages/report-server/src/routes/helpers.ts:
```diff
@@ -30,7 +28,7 @@ export const getAccessibleOrgUnitCodes = async (
   hierarchy: string,
   orgUnitCodes: string | string[],
-  entityApi: EntityApi,
+  entityApi: TupaiaApiClient['entity'],
```
Very clean 👍

kael89 commented:
@rohan-bes squashed and merged in the base branch, let's remember to mention this card in the merge commit and release PR when `rn-83` is merged in dev
